### ISSUE TITLE
Losses and BLEU in Tensorboard

### ISSUE THREAD

**Issue by ghost** [google/seq2seq#187]
Hi

Just for the sake of clarity I was wondering as to what the loss and BLEU you can see in Tensorboard refer to? Are these the loss and BLEU on the training set or on the evaluation/development set? Furthermore, is the loss you can see every 100 steps the loss on the training set or the evaluation/development set?

The reason I am asking, is because I figured the samples you see after every evaluation, are samples from the training set and not from the evaluation/development set. 

Kind regards

**Comment by zhyack** [comment]
Same question here!

I run the sample configuration with my own data. It shows that the model is trained well and gets a BLEU score as 23.81 on the tensorboard. But when I try to check that result, I find that the BLEU on the tensorboard is neither the same as BLEU evaluation on the train set(14.27), nor same as that on the dev set(13.24). A little upset, though, I wonder what's on the tensorboard!

By the way, can anyone tell me where to find the source code of this part? I'm new here, longing for a quick guide, thx!

**Comment by sld** [comment]
@zhyack I have the same issue now (BLEU in tensorboard and in bin.infer is very different)

**Comment by zhyack** [comment]
After several days' observation, now I think the BLEU in tensorboard is the evaluation on samples of development set.(When I change the dev set, the score changes respectively)
But why the evaluation of samples is quite better than the evaluation on the whole dev set?

**Comment by sld** [comment]
@zhyack https://github.com/google/seq2seq/blob/master/seq2seq/metrics/bleu.py#L62 - its a file where bleu calculation performed on train stage.
I tried to dump hypothesis and reference files into tmp dir and check number of lines. 
I see that evaluation performed on the whole dev set.

I think there is some issue with bin.infer.

This is my evaluation script:

```
python -m bin.infer \
  --tasks "
    - class: DecodeText" \
  --model_dir $MODEL_DIR \
  --checkpoint_path $MODEL_DIR/model.ckpt-12001 \
  --batch_size 256 \
  --input_pipeline "
    class: ParallelTextInputPipeline
    params:
      source_files:
        - $DEV_SOURCES" \
  > ${PRED_DIR}/predictions.txt


./seq2seq/bin/tools/multi-bleu.perl ${DEV_TARGETS_REF} < ${PRED_DIR}/predictions.txt
```

**Comment by sld** [comment]
It is really mysterious why bleu in train evaluation and in infer are so different (3.02 vs 2.26).
@dennybritz Could you please give some comments why this may happen?

**Comment by sld** [comment]
https://github.com/google/seq2seq/issues/214

**Comment by ghost** [comment]
The evaluation on the samples should be better in general, I think, because they come from the training data. The system has been training on them and is able to translate them after some time since it sees them over and over.

The evaluation on the dev set should be lower because it's unseen data, used to get an estimate of how the system would perform on new data. 

As to why your BLEU is different: maybe one of them uses greedy search when doing inference, while the other one uses some kind of beam search?

(just suggestions, I have no idea if it helps)

**Comment by SwordYork** [comment]
Hi, the evaluation on the samples should be better in general.
Because it is guided by the true words, for example, when we expect 'tf is great!', but we output 'theano' as the first word. During evaluation, 'tf' is feedback to the network, however, during infer, 'theano' is feedback to the network. Please refer to [estimator](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/learn/python/learn/estimators/estimator.py#L820).
